File: js-src/utils/text.js
import { present } from './utils'

var textSubstitution = function(s, bw, addCompareArea) {
    var y = bw.get('year.name'),
        fy = bw.get('indicator.firstYear.name'),
        yem = bw.get('indicator.yearEndMonth'),
        su = bw.get('indicator.subject'),
        a = bw.get('area.name'),
        sa = bw.get('area.name'),
        f = bw.get('feature.name'),
        fl = f || bw.get('indicator.topFeatureLabel'),
        fp = f ? bw.get('feature.parent') : "",
        d = bw.get('detail.name'),
        dl = d || bw.get('indicator.topDetailLabel'),
        ft = bw.get('indicator.featureText'),
        p = bw.get("prevYear"),
        ca = bw.get("compareArea.name");
    if (present(addCompareArea) && present(ca)) {
        a = "<span class='active'>" + a + "</span><span class='compare'> (and " + ca + ")</span>";
    }
    if (ft) {
        var fs = bw.get('feature.id');
        fl = ft[fs];
    }
    if (f) {
        s = s.replace(/[\[\]]/g, '');
    } else {
        s = s.replace(/\[.+?\]/g, '');
    }
    if (d) {
        s = s.replace(/[\{\}]/g, '');
    } else {
        s = s.replace(/\s?\{.+?\}/g, '');
    }
    s = s.replace(/\$year\$/g, y)
        .replace(/\$firstYear\$/g, fy)
        .replace(/\$yearEndMonth\$/g, yem)
        .replace(/\$subject\$/g, su)
        .replace(/\$area\$/g, a)
        .replace(/\$selectedArea\$/g, sa)
        .replace(/\$compareArea\$/g, ca)
        .replace(/\$prevYear\$/g, p)
        .replace(/\$feature\$/g, fl)
        .replace(/\$featureType\$/g, fp)
        .replace(/\$detail\$/g, dl);
    return s.trim();
}

var label = function(transform) {
    let transforms = ["absolute", "indexed", "percapita"];
    
    if (transforms.indexOf(transform) === -1) {
        if (transform === "absolute") {
            transform = "original";
        } else if (transform === "regionalPercentage") {
            transform = "feature-percentage";
        } else if (transform === "rate") {
            transform = "annual-rate";
        } else {
            return "";
        }
    }

    var l = textSubstitution(this.get("indicator.labels")[transform], this);
    return l[0].toUpperCase() + l.slice(1);
};

export { textSubstitution, label }

File: src/Bailiwick/View/Text.hs
{-# LANGUAGE OverloadedStrings #-}
module Bailiwick.View.Text
  ( textSubstitution
  , capitalize
  )
where


import Control.Applicative ((<|>))
import Control.Monad (join)
import Data.Maybe (isJust)
import Data.Char (isSpace, toUpper)

import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.HashMap.Strict.InsOrd as M (lookup)

import Bailiwick.Types

capitalize :: Text -> Text
capitalize inp =
  case T.uncons inp of
    Nothing -> ""
    Just (first, rest) -> T.cons (toUpper first) rest


textSubstitution
  :: Maybe Area
  -> Maybe Area
  -> Maybe Indicator
  -> Maybe FeatureId
  -> Maybe DetailId
  -> Maybe Year
  -> Text
  -> Text
textSubstitution area compareArea indicator feature detail year =
    let fy = indicatorFirstYear <$> indicator
        yem = join $ indicatorYearEndMonth <$> indicator
        indid = unIndicatorId . indicatorId <$> indicator
        sa = maybe "New Zealand" areaName area
        aid = maybe "new-zealand" areaId area
        f = featureIdText <$> feature
        fp = Nothing -- TODO feature types (Tourism spend)
        dl = Nothing -- TODO d <|> (indicatorTopDetailLabel =<< indicator)
        ip = join $ indicatorPeriod <$> indicator
        p = (-) <$> year <*> ip
        a = case (areaName <$> compareArea) of
              Just ca' ->
                     "<span class='active'>" <> sa <>
                     "</span><span class='compare'> (and " <> ca' <> ")</span>"
              _ -> sa
        fl = case indicatorFeatureText <$> indicator of
              Just (Just ft) -> (`M.lookup` ft) =<< feature
              _                -> f <|> (join (indicatorDefaultFeature <$> indicator))
        replace findStr (Just replaceStr) = T.replace findStr replaceStr
        replace _ _ = id
    in T.strip
      . replace "$indid$" indid
      . replace "$year$" (T.pack . show <$> year)
      . replace "$firstYear$" fy
      . replace "$yearEndMonth$" yem
      . T.replace "$area$" a
      . T.replace "$areaid$" aid
      . T.replace "$selectedArea$" sa
      . replace "$compareArea$" (areaName <$> compareArea)
      . replace "$compareAreaId$" (areaId <$> compareArea)
      . replace "$prevYear$" (T.pack . show <$> p)
      . replace "$feature$" fl
      . replace "$featureType$" fp
      . replace "$detail$" dl
      . T.pack
      . removeDetailBrackets detail
      . removeFeatureBrackets (f)
      . T.unpack


removeFeatureBrackets :: Maybe Text -> String -> String
removeFeatureBrackets feature =
  if isJust feature
    then filter (\c -> c /= '[' && c /= ']')
    else go
  where
    go :: String -> String
    go "" = ""
    go ('[':xs) = go . drop 1 $ dropWhile (/= ']') xs
    go (x:xs) = x:go xs

removeDetailBrackets :: Maybe Text -> String -> String
removeDetailBrackets detail =
  if isJust detail
    then filter (\c -> c /= '{' && c /= '}')
    else go
  where
    go :: String -> String
    go "" = ""
    go ('{':xs) = go . drop 1 $ dropWhile (/= '}') xs
    go (s:'{':xs) | isSpace s = go . drop 1 $ dropWhile (/= '}') xs
    go (x:xs) = x:go xs

